### ISSUE TITLE
etcd 3.4.0 server expects client cert when client-cert-auth = false

### ISSUE THREAD

**Issue by ae6rt** [etcd-io/etcd#11124]
For a single node etcd 3.4.0 server:

## Server version

```
etcd Version: 3.4.0
Git SHA: 898bd1351
Go Version: go1.12.9
Go OS/Arch: linux/amd64
```

## Client version

```
etcdctl version: 3.4.0
API version: 3.4
```

## Expected result

When client-cert-auth is false, a client need not present a certificate to the server.

(N.B. The expected result is the observed result for etcd 3.3.15).

## Observed result

The server expects a client cert

```
etcdctl --cacert=cabundle.pem --endpoints=192.168.99.1:2379  member list
{"level":"warn","ts":"2019-09-05T09:01:47.701-0700","caller":"clientv3/retry_interceptor.go:61","msg":"retrying of unary invoker failed","target":"endpoint://client-f5825ff3-002f-4f77-976c-a506dfdeef15/192.168.99.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest connection error: connection error: desc = \"transport: authentication handshake failed: remote error: tls: bad certificate\""}
Error: context deadline exceeded
```

The server logs accordingly

```
{"level":"info","ts":"2019-09-05T15:32:25.817Z","caller":"embed/etcd.go:708","msg":"starting with client TLS","tls-info":"cert = /tls/server.pem, key = /tls/server-key.pem, trusted-ca = /tls/ca.pem, client-cert-auth = false, crl-file = ","cipher-suites":[]}
...
{"level":"warn","ts":"2019-09-05T16:01:45.237Z","caller":"embed/config_logging.go:279","msg":"rejected connection","remote-addr":"172.17.0.1:46882","server-name":"","error":"tls: client didn't provide a certificate"}
```

When the client presents a certificate, the operation succeeds

```
etcdctl  --cacert ~/lib/easycert-ca/cabundle.pem --endpoints=192.168.99.1:2379  --cert client-cert.pem --key client-key.pem  member list
521b45b49ab8a3b9, started, 49c0b6d3b201, https://49c0b6d3b201:2380, https://49c0b6d3b201:2379, false
```

## Server configuration

```
readonly HOSTN=$(hostname)

# Establish defaults
export ETCD_NAME=${ETCD_NAME:-${HOSTN}}
export ETCD_DATA_DIR=${ETCD_DATA_DIR:-/etc-data}
export ETCD_WAL_DIR=${ETCD_WAL_DIR:-/etc-wal}
export ETCD_INITIAL_CLUSTER=${ETCD_INITIAL_CLUSTER:-${HOSTN}=https://${HOSTN}:2380}
export ETCD_INITIAL_CLUSTER_TOKEN=${ETCD_INITIAL_CLUSTER_TOKEN:-tkn}
export ETCD_AUTO_COMPACTION_RETENTION=${ETCD_AUTO_COMPACTION_RETENTION:-1} # 1 hour
export ETCD_QUOTA_BACKEND_BYTES=${ETCD_QUOTA_BACKEND_BYTES:-2147483648} # 2Gb - machine should have at least this much RAM

exec /opt/etcd/etcd \
	--listen-client-urls=https://0.0.0.0:2379 \
	--advertise-client-urls=https://${HOSTN}:2379 \
	--listen-peer-urls=https://0.0.0.0:2380 \
	--initial-advertise-peer-urls=https://${HOSTN}:2380 \
	--initial-cluster-state=new \
	--cert-file=/tls/server.pem \
	--key-file=/tls/server-key.pem \
	--trusted-ca-file=/tls/ca.pem \
	--peer-cert-file=/tls/server.pem \
	--peer-key-file=/tls/server-key.pem \
	--peer-trusted-ca-file=/tls/ca.pem \
	--logger=zap
```

**Comment by mygrsun** [comment]
if you set the 'trusted-ca-file' ,server will automatically open client check。
you can see in the code below。set ’trusted-ca-file‘ or ’client-cert-auth‘ has the same effect。
so if you don't want to check ,just remove the '--trusted-ca-file=/tls/ca.pem '

https://github.com/etcd-io/etcd/blob/master/pkg/transport/listener.go
ServerConfig function
`
if info.TrustedCAFile != "" || info.ClientCertAuth {

	cfg.ClientAuth = tls.RequireAndVerifyClientCert
}
`

`
// ServerConfig generates a tls.Config object for use by an HTTP server.
func (info TLSInfo) ServerConfig() (*tls.Config, error) {
	cfg, err := info.baseConfig()
	if err != nil {
		return nil, err
	}

	cfg.ClientAuth = tls.NoClientCert
	if info.TrustedCAFile != "" || info.ClientCertAuth {
		cfg.ClientAuth = tls.RequireAndVerifyClientCert
	}

	cs := info.cafiles()
	if len(cs) > 0 {
		cp, err := tlsutil.NewCertPool(cs)
		if err != nil {
			return nil, err
		}
		cfg.ClientCAs = cp
	}

	// "h2" NextProtos is necessary for enabling HTTP2 for go's HTTP server
	cfg.NextProtos = []string{"h2"}

	return cfg, nil
}

`

**Comment by ae6rt** [comment]
Thank you.  

The `trusted-ca-file` option should say that setting it is equivalent to setting `--client-cert-auth=true'.

**Comment by 2rs2ts** [comment]
This was not the case in 3.3, and it's not documented

It's definitely a bug, not a feature, and should be reverted

**Comment by WilliamDEdwards** [comment]
This issue is still relevant; a huge gotcha when upgrading from 3.3 is not 3.4.

**Comment by jmhbnz** [comment]
> This issue is still relevant; a huge gotcha when upgrading from 3.3 is not 3.4.

Thanks @WilliamDEdwards for the update. I've re-opened to enable some further discussion, is someone able to re-create the behavior using the latest 3.4.x release?

I see we had a pull request proposed to address the issue, we can revive that if somone can verify the issue on the latest 3.4.x release as mentioned above.

**Comment by WilliamDEdwards** [comment]
> > This issue is still relevant; a huge gotcha when upgrading from 3.3 is not 3.4.
> 
> Thanks @WilliamDEdwards for the update. I've re-opened to enable some further discussion. Is someone able to re-create the behavior using the latest 3.4.x release?

Yes, I can reproduce the issue.

I just upgraded a test instance from 3.3 to 3.4. Hence my previous comment.

One could argue setting `trusted-ca-file` without `client-cert-auth` doesn't make sense. But implicitly setting an option from one version to the next - without documentation - is probably not the way to go :)

> 
> I see we had a pull request proposed to address the issue, we can revive that if somone can verify the issue on the latest 3.4.x release as mentioned above.

I haven't tested the PR. But from the looks of it, it would solve the issue.

**Comment by jmhbnz** [comment]
Update: I proposed the code fix in https://github.com/etcd-io/etcd/pull/16585 however the current behavior has been present for a very long time and it was decided a documentation update was the preferred way to handle this.

Documentation fix to clarify this behaviour in our 3.4 upgrade docs and etcd configuration docs has been merged in https://github.com/etcd-io/website/pull/724 so I will close this now.
